crypto/dh/dh_lib.c

```diff
@@ -219,18 +219,6 @@ int DH_set0_pqg(DH *dh, BIGNUM *p, BIGNUM *q, BIGNUM *g)
 
     ossl_ffc_params_set0_pqg(&dh->params, p, q, g);
     dh_cache_named_group(dh);
-    if (q != NULL)
```

Comment:
This was here for a reason. Please search for every instance of ->length before making impulsive decisions like this.
e.g see dh_check_priv_key().

Reply:
Ok.
So I guess the code you have deleted was equivalent to bounding the length - even if it was doing it in a dodgy way.

It may still need upper bound checks on the length whenever we change q or the private key maybe??

Reply:
> It may still need upper bound checks on the length whenever we change q or the private key maybe??

This is the wrong place to do so, and a read of SP 800-56Ar3 agrees with me.  It does specify conditions on *N* during key generation, in 5.6.1.13 and 5.6.1.1.4, and I do believe we codify that (possibly partly) here:

https://github.com/openssl/openssl/blob/f4bd510503071e20472032dc9d053810a1937e95/crypto/dh/dh_key.c#L259-L261

For PKCS#3, it seems that we use a default when the length is unset:

https://github.com/openssl/openssl/blob/f4bd510503071e20472032dc9d053810a1937e95/crypto/dh/dh_key.c#L273

That does seem a bit sloppy, considering that [PKCS#3](https://www.teletrust.de/fileadmin/files/oid/oid_pkcs-3v1-4.pdf) says this:

> The central authority may optionally select an integer _l_, the private-value length in bits, that satisfies 2<sup><i>l</i>-1</sup>≤<i>p</i>.